### PR TITLE
Support for FeedOutOfDateTrigger in VulnerabilitiesGate for Grype provider 

### LINKED ISSUE #1180: Grype Scanner does not support FeedOutOfDateTrigger
FeedOutOfDateTrigger should fire for the grype DB when the active DB in use was generated more than the rule-specified number of days ago.

The v1 behavior of OS-specific stuff is not necessary given the changed semantics of how the Grype DB is built. Increasingly, vuln matches for a given package type come from multiple sources, so a single feed is no longer the right check for this trigger.

 

The v1 behavior should not be changed, just a new check implemented for the v2 scanner.

 

Steps:

    Implement trigger for grype (or variation of existing trigger). Implementation specific logic should be relegated to provider.

    Write unit tests

### PR DESCRIPTION
**Pull request by Vijay-P**
**DO NOT BE ALARMED** by large diff - pr includes a simple refactor of time utilities, and should be reviewed commit-by-commit

**What this PR does / why we need it**:
This PR adds support to the FeedOutOfDateTrigger in VulnerabilitiesGate for the Grype provider. This change will enable the trigger to fire when the grype feed data is stale.

* add support for Grype provider in VulnerabilitiesGate.FeedOutOfDateTrigger
* creates a new module and class for gate specific provider logic called `gate_util_provider.GateUtilProvider`. Hopefully, this will enhance readability and testability while avoiding overloading the VulnerabilityProvider interface.
* update VulnerabilitiesGate unit tests to include grype in test for FeedOutOfDateTrigger
* Adds unit test for  GateUtilProvider.oldest_namespace_feed_sync
* refactor time methods from anchore_engine.utils to new module, anchore_engine.util.time
* add type hints to refactored time methods
* renames VulnerabilityProvider.get_sync_utils to VulnerabilityProvider.get_sync_util_provider

**Which issue this PR fixes** *(optional, in `fixes #<issue number>)(, fixes #<issue_number, ...)` format, will close the issue when PR is merged*:
fixes #1180

**Special notes**:
This PR is based off of the following fork/branch written by @zhill: https://github.com/zhill/anchore-engine/commit/966365a9dee4fdbe68f135f641d72fe86d8619e9

Note: I could have added tests for the rest of the VulnerabilitesGate for grype, but that seemed out of scope for this ticket.

### CODE REVIEW

File: anchore_engine/services/policy_engine/engine/policy/gate_util_provider.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    get_most_recent_active_grypedb,
+)
+from anchore_engine.services.policy_engine.engine.feeds.db import (
+    get_feed_group_detached,
+)
+from anchore_engine.services.policy_engine.engine.feeds.feeds import feed_registry
+from anchore_engine.subsys import logger
+
+
+class GateUtilProvider(ABC):
```
Comment (dspalmer99) [inline anchore_engine/services/policy_engine/engine/policy/gate_util_provider.py:17]: Nice, I like the new abstraction.

File: anchore_engine/services/policy_engine/engine/policy/gate_util_provider.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    # If you need information from the Provider in the future, pass it in via the constructor in VulnerabilityProvider.get_gate_utils().
```
Comment (dspalmer99) [inline anchore_engine/services/policy_engine/engine/policy/gate_util_provider.py:18]: Could you change this comment into a docstring, and add a bit more text to be more specific about what this class is intended to encapsulate?

**Review by dspalmer99 — COMMENTED** [review]
Nice, LGTM, one ask for additional documentation.

File: anchore_engine/services/policy_engine/engine/policy/gate_util_provider.py (hunk 2 above)
Comment (Vijay-P) [reply anchore_engine/services/policy_engine/engine/policy/gate_util_provider.py]: done!

File: anchore_engine/services/policy_engine/engine/vulns/providers.py [hunk 3]
```diff
@@ -266,6 +281,13 @@ def delete_image_vulnerabilities(self, image: Image, db_session):
         """
         ...
 
+    @abstractmethod
+    def get_gate_utils(self) -> GateUtilProvider:
```
Comment (Vijay-P) [inline anchore_engine/services/policy_engine/engine/vulns/providers.py:285]: Not sure if i should change this to `get_gate_util_provider()`. There is another method in this ABC that I created previously called `get_sync_utils()`, so I was following that naming convention.

Comment (Toure) [reply anchore_engine/services/policy_engine/engine/vulns/providers.py:285]: I like the get_gate_util_provider as the name is more explicit.

**Review by Toure — COMMENTED** [review]
This looks pretty good, will give it a spin before an ack

File: anchore_engine/services/policy_engine/engine/vulns/providers.py (hunk 3 above)
Comment (Vijay-P) [reply anchore_engine/services/policy_engine/engine/vulns/providers.py:285]: Ok. Going to rename both. Will add the rename for get_gate_utils as a fixup in the first commit. Will add a new commit for renaming get_sync_utils.

Comment (rbrady) [reply anchore_engine/services/policy_engine/engine/vulns/providers.py]: I am also +1 for naming to be more explicit.  The plurality of the method names would be more suited to returning a collection of something.   I would also change the "get_sync_utils" method to "get_sync_util_provider".

**Review by rbrady — COMMENTED** [review]
Overall a good approach.  Couple of nits for naming.

**Review by rbrady — APPROVED** [review]
Thanks for the updates.  LGTM.  Thanks for the improvements.

**Pull request merged**